File: packages/ai-parrot-server/ui/src/lib/components/agents/canvas/a2ui/linked/conditions.ts
/**
 * Canonical request → QuerySource conditions (FEAT-598, spec §7 / S5).
 * TS twin of `parrot.outputs.a2ui.linked.conditions.derive_conditions`; pinned by
 * `contract/fixtures/conditions/*.json`. Never emits `querylimit` or `refresh` (lane-time keys).
 */
import type { SourceRequest } from './types';

// Keys a lane adds at fetch time; derive_conditions never emits them.
const LANE_TIME_KEYS: Set<string> = new Set(['querylimit', 'refresh']);

export function deriveConditions(
  request: SourceRequest,
  locked: Record<string, unknown>,
): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  
  // Rule 1: placeholders in request order, locked values overriding same-named keys
  // (locked-only keys appended after, as Python does)
  for (const [key, value] of Object.entries(request.placeholders)) {
    // Never emit lane-time keys
    if (!LANE_TIME_KEYS.has(key)) {
      out[key] = value;
    }
  }
  for (const [key, value] of Object.entries(locked)) {
    // Never emit lane-time keys
    if (!LANE_TIME_KEYS.has(key)) {
      out[key] = value;
    }
  }
  
  // Rule 2: filter entries verbatim
  if (request.filter && Object.keys(request.filter).length > 0) {
    out['filter'] = { ...request.filter };
  }
  
  // Rule 3: fields / ordering / grouping only when non-empty
  if (request.fields && request.fields.length > 0) {
    out['fields'] = [...request.fields];
  }
  if (request.ordering && request.ordering.length > 0) {
    out['ordering'] = [...request.ordering];
  }
  if (request.grouping && request.grouping.length > 0) {
    out['grouping'] = [...request.grouping];
  }
  
  // Rule 4: limit / offset mapped to the same dialect keys the fixtures use
  if (request.limit !== undefined && request.limit !== null) {
    out['limit'] = request.limit;
  }
  if (request.offset !== undefined && request.offset !== null) {
    out['_offset'] = request.offset;
  }
  
  return out;
}

File: packages/ai-parrot/src/parrot/outputs/a2ui/linked/conditions.py
"""Canonical ``SourceRequest`` → QuerySource ``conditions`` derivation (FEAT-598 S5, spec §7).

Every executor re-implements these rules from ``contract/fixtures/conditions/*.json``.
"""
from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from parrot.outputs.a2ui.linked.models import LinkedDataSource, SourceRequest

#: Keys a lane adds at fetch time; derive_conditions never emits them.
LANE_TIME_KEYS: frozenset[str] = frozenset({"querylimit", "refresh"})


def derive_conditions(request: SourceRequest, *, locked: Mapping[str, Any]) -> dict[str, Any]:
    """Return the deterministic QuerySource payload for ``request`` with ``locked`` values applied.

    Args:
        request: The structured, editable request.
        locked: Locked parameter values; they win over same-named placeholders.

    Returns:
        A new dict; key order is part of the contract.
    """
    payload: dict[str, Any] = dict(request.placeholders)
    for key, value in locked.items():
        payload[key] = value
    if request.filter:
        payload["filter"] = dict(request.filter)
    for key, val in (("fields", request.fields), ("ordering", request.ordering), ("grouping", request.grouping)):
        if val:
            payload[key] = list(val)
    if request.offset:
        payload["_offset"] = int(request.offset)
    return payload


def locked_values(source: LinkedDataSource) -> dict[str, Any]:
    """Return ``{name: value}`` for every locked name present in ``source.conditions``."""
    return {name: source.conditions[name] for name in source.locked if name in source.conditions}
